drivers/axi_core/axi_sysid/axi_sysid.c

```diff
@@ -0,0 +1,297 @@
+		build->git_hash,
+		(build->git_clean_chk[0] == 't') ? "clean" : "dirty");
+
+	return 0;
+}
+
+static int axi_sysid_validate(struct axi_sysid *st)
+{
+	struct build_info_header_v1_1 *build_v1_1;
+	struct build_info_header_v1 *build_v1;
+	struct sysid_header_v1 *header;
+	uint32_t *build_raw;
+
+	st->mem = no_os_calloc(st->size, sizeof(uint32_t));
+	if (!st->mem)
+		return -ENOMEM;
+
+	axi_sysid_read(st, 0, st->mem, st->size);
+	header = (struct sysid_header_v1 *) st->mem;
+
+	if (axi_sysid_checksum((uint8_t *)st->mem, sizeof(struct sysid_header_v1))) {
+		pr_err("verify header checksum failed\n");
+		return -EFAULT;
+	}
+
+	if (header->version != st->info->header_version) {
+		pr_err("system ID header version mismatch. Expected %d, Reported %d\n",
+		       st->info->header_version, header->version);
+		return -EFAULT;
+	}
+
+	build_raw = &st->mem[header->build_info_offs];
+
+	switch (header->version) {
+	case AXI_SYSID_HEADER_V2:
+		build_v1_1 = (struct build_info_header_v1_1 *) build_raw;
+
+		return axi_sysid_validate_v1_1(st, build_v1_1);
+	case AXI_SYSID_HEADER_V1:
+		build_v1 = (struct build_info_header_v1 *) build_raw;
+
+		return axi_sysid_validate_v1(st, build_v1);
+	default:
+		return -EINVAL;
+	}
+}
+
+/*******************************************************************************
+* @brief Initialization function for AXI_SYID.
+*
+* @param sysid_core - SYSID istance.
+* @param init - Structure containing initializing data.
+*
+* @return 0 for success, -1 in case of failure.
+*******************************************************************************/
+int32_t axi_sysid_init(struct axi_sysid **sysid_core,
+		       const struct axi_sysid_init_param *init)
+{
+	struct axi_sysid_core_info *info;
+	struct axi_sysid *sysid;
+	int ret;
+
+	sysid = (struct axi_sysid *)no_os_calloc(1, sizeof(*sysid));
+	if (!sysid)
+		return -1;
+
+	sysid->base = init->base;
+	sysid->size = (1 << axi_sysid_ioread(sysid, AXI_SYSID_REG_ROM_ADDR_WIDTH)) *
+		      AXI_SYSID_WORD_SIZE;
+
+	info = (struct axi_sysid_core_info *)no_os_calloc(1, sizeof(*info));
+	if (!info)
+		goto error;
+
+	info->version = axi_sysid_ioread(sysid, AXI_SYSID_REG_VERSION);
+
+	sysid->info = info;
+
+	switch(sysid->info->version) {
+	case 0x10061:
+		sysid->info->header_version = AXI_SYSID_HEADER_V1;
+		break;
+	case 0x10161:
+		sysid->info->header_version = AXI_SYSID_HEADER_V2;
+		break;
+	}
+
+	ret = axi_sysid_validate(sysid);
+	if (ret)
+		goto error1;
+
+	*sysid_core = sysid;
+
+	return 0;
+error1:
+	no_os_free(info);
+error:
+	no_os_free(sysid);
+	return -1;
+}
+
+/*******************************************************************************
+* @brief Remove the SYSID instance.
+*
+* @param sysid - SYSID istance.
+*
+* @return 0 for success, -1 in case of failure.
+*******************************************************************************/
+int32_t axi_sysid_remove(struct axi_sysid *sysid)
+{
+	if (!sysid)
+		return -1;
```

Comment:
ditto.

Reply:
Thank you for the comments. The code has been updated accordingly.